<a href="https://colab.research.google.com/github/vincentjunitio00/Iris-Flower/blob/master/Iris_Flower_Dataset_Notebook.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Iris Flower Dataset Notebook #
#### by: Vincent Junitio Ungu ####

The dataset is taken from Iris Flower Dataset at Kaggle. Link: https://www.kaggle.com/arshid/iris-flower-dataset


This dataset is used to do a multi-class classification (3 classes). It only contains 150 records. The attributes in the set are:

* Petal Length: the length of the petal - modified leaf which surround the reproductive parts of a flower.
* Petal Width: the width of the petal.
* Sepal Length: the length of the sepal - the outermost part of a flower which encloses its developing reproductive structures.
* Sepal Width: the width of the sepal.
* Species: the species of the flower (Iris setosa, Iris versicolor or Iris virginica)

For those of you who love to start learning data science and machine learning, this notebook will be a small kick start for you to dive in those world. :) Enjoy it!

First, let's import the csv file from Kaggle. I have downloaded the file and it is in my repo, so I will import it from my github repo. 

In [0]:
import pandas as pd
iris = pd.read_csv('https://raw.githubusercontent.com/vincentjunitio00/Iris-Flower/master/Iris.csv')

Let's see how is the dataset looks like.

In [599]:
iris

Id  SepalLengthCm  ...  PetalWidthCm         Species
0      1            5.1  ...           0.2     Iris-setosa
1      2            4.9  ...           0.2     Iris-setosa
2      3            4.7  ...           0.2     Iris-setosa
3      4            4.6  ...           0.2     Iris-setosa
4      5            5.0  ...           0.2     Iris-setosa
..   ...            ...  ...           ...             ...
145  146            6.7  ...           2.3  Iris-virginica
146  147            6.3  ...           1.9  Iris-virginica
147  148            6.5  ...           2.0  Iris-virginica
148  149            6.2  ...           2.3  Iris-virginica
149  150            5.9  ...           1.8  Iris-virginica

[150 rows x 6 columns]

I notice that the dataset is somehow sorted by the "Species" column. Let's have a check is it really true or not.

In [600]:
iris[iris['Species']=='Iris-setosa']

Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm      Species
0    1            5.1           3.5            1.4           0.2  Iris-setosa
1    2            4.9           3.0            1.4           0.2  Iris-setosa
2    3            4.7           3.2            1.3           0.2  Iris-setosa
3    4            4.6           3.1            1.5           0.2  Iris-setosa
4    5            5.0           3.6            1.4           0.2  Iris-setosa
5    6            5.4           3.9            1.7           0.4  Iris-setosa
6    7            4.6           3.4            1.4           0.3  Iris-setosa
7    8            5.0           3.4            1.5           0.2  Iris-setosa
8    9            4.4           2.9            1.4           0.2  Iris-setosa
9   10            4.9           3.1            1.5           0.1  Iris-setosa
10  11            5.4           3.7            1.5           0.2  Iris-setosa
11  12            4.8           3.4            1.6           0.2  Iris-setosa
12  13            4.8           3.0            1.4           0.1  Iris-setosa
13  14            4.3           3.0            1.1           0.1  Iris-setosa
14  15            5.8           4.0            1.2           0.2  Iris-setosa
15  16            5.7           4.4            1.5           0.4  Iris-setosa
16  17            5.4           3.9            1.3           0.4  Iris-setosa
17  18            5.1           3.5            1.4           0.3  Iris-setosa
18  19            5.7           3.8            1.7           0.3  Iris-setosa
19  20            5.1           3.8            1.5           0.3  Iris-setosa
20  21            5.4           3.4            1.7           0.2  Iris-setosa
21  22            5.1           3.7            1.5           0.4  Iris-setosa
22  23            4.6           3.6            1.0           0.2  Iris-setosa
23  24            5.1           3.3            1.7           0.5  Iris-setosa
24  25            4.8           3.4            1.9           0.2  Iris-setosa
25  26            5.0           3.0            1.6           0.2  Iris-setosa
26  27            5.0           3.4            1.6           0.4  Iris-setosa
27  28            5.2           3.5            1.5           0.2  Iris-setosa
28  29            5.2           3.4            1.4           0.2  Iris-setosa
29  30            4.7           3.2            1.6           0.2  Iris-setosa
30  31            4.8           3.1            1.6           0.2  Iris-setosa
31  32            5.4           3.4            1.5           0.4  Iris-setosa
32  33            5.2           4.1            1.5           0.1  Iris-setosa
33  34            5.5           4.2            1.4           0.2  Iris-setosa
34  35            4.9           3.1            1.5           0.1  Iris-setosa
35  36            5.0           3.2            1.2           0.2  Iris-setosa
36  37            5.5           3.5            1.3           0.2  Iris-setosa
37  38            4.9           3.1            1.5           0.1  Iris-setosa
38  39            4.4           3.0            1.3           0.2  Iris-setosa
39  40            5.1           3.4            1.5           0.2  Iris-setosa
40  41            5.0           3.5            1.3           0.3  Iris-setosa
41  42            4.5           2.3            1.3           0.3  Iris-setosa
42  43            4.4           3.2            1.3           0.2  Iris-setosa
43  44            5.0           3.5            1.6           0.6  Iris-setosa
44  45            5.1           3.8            1.9           0.4  Iris-setosa
45  46            4.8           3.0            1.4           0.3  Iris-setosa
46  47            5.1           3.8            1.6           0.2  Iris-setosa
47  48            4.6           3.2            1.4           0.2  Iris-setosa
48  49            5.3           3.7            1.5           0.2  Iris-setosa
49  50            5.0           3.3            1.4           0.2  Iris-setosa

In [601]:
iris[iris['Species']=='Iris-versicolor']

Id  SepalLengthCm  ...  PetalWidthCm          Species
50   51            7.0  ...           1.4  Iris-versicolor
51   52            6.4  ...           1.5  Iris-versicolor
52   53            6.9  ...           1.5  Iris-versicolor
53   54            5.5  ...           1.3  Iris-versicolor
54   55            6.5  ...           1.5  Iris-versicolor
55   56            5.7  ...           1.3  Iris-versicolor
56   57            6.3  ...           1.6  Iris-versicolor
57   58            4.9  ...           1.0  Iris-versicolor
58   59            6.6  ...           1.3  Iris-versicolor
59   60            5.2  ...           1.4  Iris-versicolor
60   61            5.0  ...           1.0  Iris-versicolor
61   62            5.9  ...           1.5  Iris-versicolor
62   63            6.0  ...           1.0  Iris-versicolor
63   64            6.1  ...           1.4  Iris-versicolor
64   65            5.6  ...           1.3  Iris-versicolor
65   66            6.7  ...           1.4  Iris-versicolor
66   67            5.6  ...           1.5  Iris-versicolor
67   68            5.8  ...           1.0  Iris-versicolor
68   69            6.2  ...           1.5  Iris-versicolor
69   70            5.6  ...           1.1  Iris-versicolor
70   71            5.9  ...           1.8  Iris-versicolor
71   72            6.1  ...           1.3  Iris-versicolor
72   73            6.3  ...           1.5  Iris-versicolor
73   74            6.1  ...           1.2  Iris-versicolor
74   75            6.4  ...           1.3  Iris-versicolor
75   76            6.6  ...           1.4  Iris-versicolor
76   77            6.8  ...           1.4  Iris-versicolor
77   78            6.7  ...           1.7  Iris-versicolor
78   79            6.0  ...           1.5  Iris-versicolor
79   80            5.7  ...           1.0  Iris-versicolor
80   81            5.5  ...           1.1  Iris-versicolor
81   82            5.5  ...           1.0  Iris-versicolor
82   83            5.8  ...           1.2  Iris-versicolor
83   84            6.0  ...           1.6  Iris-versicolor
84   85            5.4  ...           1.5  Iris-versicolor
85   86            6.0  ...           1.6  Iris-versicolor
86   87            6.7  ...           1.5  Iris-versicolor
87   88            6.3  ...           1.3  Iris-versicolor
88   89            5.6  ...           1.3  Iris-versicolor
89   90            5.5  ...           1.3  Iris-versicolor
90   91            5.5  ...           1.2  Iris-versicolor
91   92            6.1  ...           1.4  Iris-versicolor
92   93            5.8  ...           1.2  Iris-versicolor
93   94            5.0  ...           1.0  Iris-versicolor
94   95            5.6  ...           1.3  Iris-versicolor
95   96            5.7  ...           1.2  Iris-versicolor
96   97            5.7  ...           1.3  Iris-versicolor
97   98            6.2  ...           1.3  Iris-versicolor
98   99            5.1  ...           1.1  Iris-versicolor
99  100            5.7  ...           1.3  Iris-versicolor

[50 rows x 6 columns]

In [602]:
iris[iris['Species']=='Iris-virginica']

Id  SepalLengthCm  ...  PetalWidthCm         Species
100  101            6.3  ...           2.5  Iris-virginica
101  102            5.8  ...           1.9  Iris-virginica
102  103            7.1  ...           2.1  Iris-virginica
103  104            6.3  ...           1.8  Iris-virginica
104  105            6.5  ...           2.2  Iris-virginica
105  106            7.6  ...           2.1  Iris-virginica
106  107            4.9  ...           1.7  Iris-virginica
107  108            7.3  ...           1.8  Iris-virginica
108  109            6.7  ...           1.8  Iris-virginica
109  110            7.2  ...           2.5  Iris-virginica
110  111            6.5  ...           2.0  Iris-virginica
111  112            6.4  ...           1.9  Iris-virginica
112  113            6.8  ...           2.1  Iris-virginica
113  114            5.7  ...           2.0  Iris-virginica
114  115            5.8  ...           2.4  Iris-virginica
115  116            6.4  ...           2.3  Iris-virginica
116  117            6.5  ...           1.8  Iris-virginica
117  118            7.7  ...           2.2  Iris-virginica
118  119            7.7  ...           2.3  Iris-virginica
119  120            6.0  ...           1.5  Iris-virginica
120  121            6.9  ...           2.3  Iris-virginica
121  122            5.6  ...           2.0  Iris-virginica
122  123            7.7  ...           2.0  Iris-virginica
123  124            6.3  ...           1.8  Iris-virginica
124  125            6.7  ...           2.1  Iris-virginica
125  126            7.2  ...           1.8  Iris-virginica
126  127            6.2  ...           1.8  Iris-virginica
127  128            6.1  ...           1.8  Iris-virginica
128  129            6.4  ...           2.1  Iris-virginica
129  130            7.2  ...           1.6  Iris-virginica
130  131            7.4  ...           1.9  Iris-virginica
131  132            7.9  ...           2.0  Iris-virginica
132  133            6.4  ...           2.2  Iris-virginica
133  134            6.3  ...           1.5  Iris-virginica
134  135            6.1  ...           1.4  Iris-virginica
135  136            7.7  ...           2.3  Iris-virginica
136  137            6.3  ...           2.4  Iris-virginica
137  138            6.4  ...           1.8  Iris-virginica
138  139            6.0  ...           1.8  Iris-virginica
139  140            6.9  ...           2.1  Iris-virginica
140  141            6.7  ...           2.4  Iris-virginica
141  142            6.9  ...           2.3  Iris-virginica
142  143            5.8  ...           1.9  Iris-virginica
143  144            6.8  ...           2.3  Iris-virginica
144  145            6.7  ...           2.5  Iris-virginica
145  146            6.7  ...           2.3  Iris-virginica
146  147            6.3  ...           1.9  Iris-virginica
147  148            6.5  ...           2.0  Iris-virginica
148  149            6.2  ...           2.3  Iris-virginica
149  150            5.9  ...           1.8  Iris-virginica

[50 rows x 6 columns]

Yes it is! Now I am going to shuffle the dataset (to prevent a clustered species since it will affect the learning process by Machine Learning). 

Do we really need to shuffle the dataset?
My answer would be yes! Shuffling the dataset will reduce the variance of the data and be more generalized so the prediction will be more accurate.

In [0]:
from sklearn.utils import shuffle
iris = shuffle(iris)

Now let's see the shuffled dataset (you may ignore the index of the dataset)

In [604]:
iris

Id  SepalLengthCm  ...  PetalWidthCm          Species
133  134            6.3  ...           1.5   Iris-virginica
131  132            7.9  ...           2.0   Iris-virginica
89    90            5.5  ...           1.3  Iris-versicolor
56    57            6.3  ...           1.6  Iris-versicolor
67    68            5.8  ...           1.0  Iris-versicolor
..   ...            ...  ...           ...              ...
60    61            5.0  ...           1.0  Iris-versicolor
144  145            6.7  ...           2.5   Iris-virginica
118  119            7.7  ...           2.3   Iris-virginica
0      1            5.1  ...           0.2      Iris-setosa
105  106            7.6  ...           2.1   Iris-virginica

[150 rows x 6 columns]

Next, let's see what is the type of each columns/ attributes.

In [605]:
iris.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 150 entries, 133 to 105
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             150 non-null    int64  
 1   SepalLengthCm  150 non-null    float64
 2   SepalWidthCm   150 non-null    float64
 3   PetalLengthCm  150 non-null    float64
 4   PetalWidthCm   150 non-null    float64
 5   Species        150 non-null    object 
dtypes: float64(4), int64(1), object(1)
memory usage: 8.2+ KB


Suppose that initially I do not know how many datas with Iris versicolor, Iris setosa and Iris virginica. I am going to count it now.

In [606]:
iris['Species'].value_counts()

Iris-versicolor    50
Iris-setosa        50
Iris-virginica     50
Name: Species, dtype: int64

Oh it's 50 datas of each species.

Since our task is to do a multi-class classification or we can say that we are going to predict whether this data belongs to Iris setosa, iris versicolor or iris virginica, I will change each value with an integer - 0 for Iris setosa, 1 for Iris versicolor, 2 for Iris virginica

In [0]:
pd.options.mode.chained_assignment = None
def changeValue(dataframe):
  dataframe['Species'][dataframe['Species']=='Iris-setosa'] = 0
  dataframe['Species'][dataframe['Species']=='Iris-versicolor'] = 1
  dataframe['Species'][dataframe['Species']=='Iris-virginica']= 2

In [0]:
changeValue(iris)

In [609]:
iris.reset_index(drop=True)

Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm Species
0    134            6.3           2.8            5.1           1.5       2
1    132            7.9           3.8            6.4           2.0       2
2     90            5.5           2.5            4.0           1.3       1
3     57            6.3           3.3            4.7           1.6       1
4     68            5.8           2.7            4.1           1.0       1
..   ...            ...           ...            ...           ...     ...
145   61            5.0           2.0            3.5           1.0       1
146  145            6.7           3.3            5.7           2.5       2
147  119            7.7           2.6            6.9           2.3       2
148    1            5.1           3.5            1.4           0.2       0
149  106            7.6           3.0            6.6           2.1       2

[150 rows x 6 columns]

As you see, the value in "Species" attribute is now changed into a number ranging from 0 to 2.

It's training time!
But before that, I have to split the dataset into training set and test set. Feel free to tweak the length but please make sure that the number of the training set is larger the number of the test set. Since I use 0.94 for the length variable, I will have 141 datas for training set and 9 datas for test set - in this case the number of set that we are going to predict is 9 datas.

In [0]:
length = 0.94
split = int(length * len(iris))
train_set, test_set = iris[:split], iris[split:]

In [611]:
train_set = train_set.reset_index(drop = True)
train_set

Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm Species
0    134            6.3           2.8            5.1           1.5       2
1    132            7.9           3.8            6.4           2.0       2
2     90            5.5           2.5            4.0           1.3       1
3     57            6.3           3.3            4.7           1.6       1
4     68            5.8           2.7            4.1           1.0       1
..   ...            ...           ...            ...           ...     ...
136    9            4.4           2.9            1.4           0.2       0
137   12            4.8           3.4            1.6           0.2       0
138  150            5.9           3.0            5.1           1.8       2
139   49            5.3           3.7            1.5           0.2       0
140    3            4.7           3.2            1.3           0.2       0

[141 rows x 6 columns]

In [612]:
test_set = test_set.reset_index(drop = True)
test_set

Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm Species
0   32            5.4           3.4            1.5           0.4       0
1   98            6.2           2.9            4.3           1.3       1
2   64            6.1           2.9            4.7           1.4       1
3  102            5.8           2.7            5.1           1.9       2
4   61            5.0           2.0            3.5           1.0       1
5  145            6.7           3.3            5.7           2.5       2
6  119            7.7           2.6            6.9           2.3       2
7    1            5.1           3.5            1.4           0.2       0
8  106            7.6           3.0            6.6           2.1       2

The "Species" attribute will be fed into y_train variable. The "Id" attribute will not give us any information about the species so I am going to drop these two attributes. (Do not be confused with the variable name, sometimes people say columns, sometimes people say attributes, sometimes people say features. I personally make columns are the same as attributes, while features are the attributes that we take into account for the learning proses)

In [0]:
drop_features = ["Id","Species"]
x_train_data = train_set.drop(drop_features,axis = 1)

In [614]:
x_train_data

SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm
0              6.3           2.8            5.1           1.5
1              7.9           3.8            6.4           2.0
2              5.5           2.5            4.0           1.3
3              6.3           3.3            4.7           1.6
4              5.8           2.7            4.1           1.0
..             ...           ...            ...           ...
136            4.4           2.9            1.4           0.2
137            4.8           3.4            1.6           0.2
138            5.9           3.0            5.1           1.8
139            5.3           3.7            1.5           0.2
140            4.7           3.2            1.3           0.2

[141 rows x 4 columns]

I will save the "Species" attribute to y_train variable.

In [0]:
y_train = train_set["Species"]

y_train should be in int64, so let's change its type.

In [0]:
y_train = y_train.astype('int64')

Import the machine learning model -> Support Vector Classification. Check this link below to learn more about Support Vector Classification. 
https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVC.html

In [0]:
from sklearn.svm import SVC

In [618]:
svc = SVC()
# fit it to start training
svc.fit(x_train_data, y_train)

SVC(C=1.0, break_ties=False, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma='scale', kernel='rbf',
    max_iter=-1, probability=False, random_state=None, shrinking=True,
    tol=0.001, verbose=False)

I will repeat the same process like what I have done with the training set to the test set (specifying the "Species" to y_true and drop unused features/ attributes to x_test_set). 

In [0]:
y_true = test_set['Species']

In [0]:
y_true = y_true.astype('int64')

The first column is the index column, while the second column is the "Species" column. Since we are using the second column only, I will convert it into a numpy array.

In [621]:
y_true

0    0
1    1
2    1
3    2
4    1
5    2
6    2
7    0
8    2
Name: Species, dtype: int64

You can see what's the difference after I have converted it into numpy array.

In [622]:
y_true_numpy = y_true.to_numpy()
y_true_numpy

array([0, 1, 1, 2, 1, 2, 2, 0, 2])

In [0]:
x_test_set = test_set.drop(["Id","Species"],axis = 1)

In [624]:
x_test_set

SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm
0            5.4           3.4            1.5           0.4
1            6.2           2.9            4.3           1.3
2            6.1           2.9            4.7           1.4
3            5.8           2.7            5.1           1.9
4            5.0           2.0            3.5           1.0
5            6.7           3.3            5.7           2.5
6            7.7           2.6            6.9           2.3
7            5.1           3.5            1.4           0.2
8            7.6           3.0            6.6           2.1

After training, it's time to predict! I am going to predict the x_test_set. Remember that the x_test_set consists of datas in which the models have not yet trained.

In [0]:
predict = svc.predict(x_test_set)

In [626]:
predict

array([0, 1, 1, 2, 1, 2, 2, 0, 2])

I use confusion matrix to see how accurate my prediction is.

In [627]:
from sklearn.metrics import confusion_matrix
confusion_matrix(y_true_numpy, predict, labels = [0,1,2])

array([[2, 0, 0],
       [0, 3, 0],
       [0, 0, 4]])

The confusion matrix above is 3x3. Suppose the matrix is in a Cartesian coordinate. The y axis is the true label, while the x axis is the predict label. The accurate prediction will have a value in the main diagonal of the confusion matrix (index = [0][0], [1][1], [2][2]). I set the parameter labels = [0,1,2] showing the order of the columns and rows in the matrix. First column and row for the 0 -> Iris-setosa. Second column and row for the 1 -> Iris-versicolor. Third column and row for the 2 -> Iris-virginica.


Feel free to play around the length variable and the machine learning techniques to see how it affects your prediction. I obtained a 100% accurate for 9 out of 10 times with this length and model.

This is just a basic and I hope you guys enjoy it. If you are interested in dive in the world of data science and machine learning after this, I would be so happy to have been provided this notebook. :))